citar-cache.el

```diff
@@ -0,0 +1,239 @@
+not included in FILENAMES.  Release cached files that are no
+longer needed by any other buffer.
+
+Return a list of `citar--bibliography` objects, one for each
+element of FILENAMES."
+  (citar-cache--release-bibliographies filenames buffer)
+  (mapcar
+   (lambda (filename) (citar-cache--get-bibliography filename buffer))
+   filenames))
+
+(defun citar-cache--entry (key bibs)
+  (catch :found
+    ;; Iterate through the cached bibliography hashes and find a key.
+    (dolist (bib bibs)
+      (let* ((entries (citar-cache--bibliography-entries bib))
+             (entry (gethash key entries)))
+        (when entry (throw :found entry))))
+    nil))
+
+(defun citar-cache--entries (bibs)
+  (citar-cache--merge-hash-tables
+   (mapcar #'citar-cache--bibliography-entries bibs)))
+
+(defun citar-cache--preformatted (bibs)
+  (citar-cache--merge-hash-tables
+   (mapcar #'citar-cache--bibliography-preformatted bibs)))
+
+
+;;; Creating and deleting bibliography caches
+
+
+(defun citar-cache--get-bibliography (filename &optional buffer)
+  "Return cached bibliography for FILENAME and associate it with BUFFER.
+If FILENAME is not already cached, read and cache it.  If BUFFER
+is nil, use the current buffer.  Otherwise, BUFFER should be a
+buffer object or name that requires the bibliography FILENAME, or
+a symbol like 'global."
+  (let* ((cached (gethash filename citar-cache--bibliographies))
+         (bib (or cached (citar-cache--make-bibliography filename)))
+         (buffer (citar-cache--canonicalize-buffer buffer))
+         (fmtstr (citar--get-template 'completion)))
+    (unless cached
+      (setf (citar-cache--bibliography-format-string bib) fmtstr)
+      (citar-cache--update-bibliography bib)
+      (puthash filename bib citar-cache--bibliographies))
+    ;; Preformat strings if format has changed
+    (unless (equal-including-properties
+             fmtstr (citar-cache--bibliography-format-string bib))
+      (setf (citar-cache--bibliography-format-string bib) fmtstr)
+      (citar-cache--preformat-bibliography bib))
+    ;; Associate buffer with this bibliography:
+    (cl-pushnew buffer (citar-cache--bibliography-buffers bib))
+    ;; Release bibliography when buffer is killed or changes major mode:
+    (unless (symbolp buffer)
+      (with-current-buffer buffer
+        (dolist (hook '(change-major-mode-hook kill-buffer-hook))
+          (add-hook hook #'citar-cache--release-bibliographies 0 'local))))
+    bib))
+
+
+(defun citar-cache--release-bibliographies (&optional keep-filenames buffer)
+  "Dissociate BUFFER from cached bibliographies.
+If BUFFER is nil, use the current buffer.  Otherwise, BUFFER
+should be a buffer object, buffer name, or a symbol like 'global.
+KEEP-FILENAMES is a list of file names that are not dissociated
+from BUFFER.
+
+Remove any bibliographies from the cache that are no longer
+needed by any other buffer."
+  (let ((buffer (citar-cache--canonicalize-buffer buffer)))
+    (maphash
```

Comment:
I used to do this, using the more generic `seq` and `map` functions wherever possible. But lately, since I've been focusing on performance, I've come to realize that unnecessary genericity is not always good.

Using `maphash` tells the compiler that you're definitely operating on a hash table, which is the case because this code is accessing our own internal variables (it would be a huge bug if we got anything but a hash table). It's implemented in C and is as close to a direct loop over hash tables as you can get. Potentially, it could even get optimized away with the new native compilation work, and the lambda function inlined (which would be a huge win).

The `seq` or `map` functions, on the other hand, must check the type of the object at runtime (necessarily, because they're not macros). The check itself is quite fast, but this means that the compiler has no visibility into what code is going to run. It also has to package up the lambda and call it as a function on every iteration.

My current approach is to directly use the built-in C functions, like `mapcar` and `maphash`, for basic access when possible. I only use the generic functions for higher-level algorithms that wouldn't make sense to implement separately for each type, or for handling data of unknown types.